crates/astria-sequencer/src/transaction/state_ext.rs

```diff
@@ -39,12 +53,53 @@ pub(crate) trait StateWriteExt: StateWrite {
     fn delete_current_source(&mut self) {
         self.object_delete(current_source());
     }
+
+    #[instrument(skip_all)]
+    fn put_transaction_index_of_action(&mut self, index: u64) {
```

Comment:
While the abstraction may be nice, I wonder how the actual implementation would look. Similar to how `TransactionContext` is constructed from the transaction itself, it seems to me that it's implied `ActionContext` would be constructed from the action itself, which I believe would be problematic because the action does not contain the context of its position within the transaction. If only the transaction has this context, then IMO it seems most fitting to include it in `TransactionContext`, but given that it has to be updated every action, that would require may more reads/writes of the `TransactionContext` object. That's why I had originally separated the index from it, but if you had a different idea for how to implement `ActionContext` I'd be open to it

Reply:
You are right - the action does not have that context, the transaction does. So indeed it should be added there.

For the ephemeral cache there are actually no reads and writes. If you look at the implementation of `object_put` (or `put_object` or whatever it's called), then `TransactionContext` is put into a `HashMap<String, Box<dyn Any>>`, with the value being downcast to a `TransactionContext`. So this is strictly cheaper to writing and reading from the non-verifiable cache.